src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,260 @@
+package ui.components.pickers;
+
+import backend.resource.TurboLabel;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to represent the state of the label picker. In addition to representing a state,
+ * it contains the logic that handles every state transition
+ */
+public class LabelPickerState {
+
+    Set<String> initialLabels;
+    List<String> addedLabels;
+    List<String> removedLabels;
+    List<String> matchedLabels;
+    List<TurboLabel> repoLabels;
+    OptionalInt currentSuggestionIndex;
+
+    @SuppressWarnings("PMD")
+    public LabelPickerState(Set<String> initialLabels, List<TurboLabel> repoLabels, String userInput) {
+        this(initialLabels, new ArrayList<>(), new ArrayList<>(), new ArrayList<>(), repoLabels,
+                OptionalInt.empty());
+        update(userInput);
+    }
+
+    private LabelPickerState(Set<String> initialLabels, List<String> addedLabels, List<String> removedLabels,
+                             List<String> matchedLabels, List<TurboLabel> repoLabels, 
+                             OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.repoLabels = repoLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    /**
+     * Updates current state based on given user input
+     * @param userInput 
+     */
+    private void update(String userInput) {
+        List<String> confirmedKeywords = getConfirmedKeywords(userInput);
+        for (String confirmedKeyword : confirmedKeywords) {
+            findAndToggleMatchingLabel(confirmedKeyword);
+        }
+
+        Optional<String> keywordInProgess = getKeywordInProgress(userInput);
+        if (keywordInProgess.isPresent()) {
+            updateMatchedLabels(keywordInProgess.get());
+        }
+    }
+
+    /**
+     * Update current state if there is exactly one matching label based on the 
+     * given keyword
+     * @param keyword
+     */
+    private void findAndToggleMatchingLabel(String keyword) {
+        if (TurboLabel.hasExactlyOneMatchedLabel(repoLabels, keyword)) {
+            String labelName =
+                    TurboLabel.getFirstMatchingTurboLabel(repoLabels, keyword).getFullName();
+            toggleLabel(labelName);
+        }
+    }
+
+    /**
+     * Updates current state based on properties of selected label
+     * labelName is case-sensitive
+     * @param labelName
+     */
+    public void toggleLabel(String labelName) {
+        if (isAnInitialLabel(labelName)) {
+            if (isARemovedLabel(labelName)) {
+                // add back initial label
```

Comment:
Ya I agree this is a bit misleading because it is added back in a sense because it is remove from the removedLabels.